### PR TITLE
Add sspBC adapter

### PR DESCRIPTION
**Pull request by wojciech-bialy-wpm**
## Type of change

- [x] New bidder adapter   

## Description of change

Add bidder adapter used by Wirtualna Polska Media  and it's partners

- test parameters for validating bids
```
   {
      code: 'banner-div',
      mediaTypes: {
        banner: {
          sizes: [[970,300],
                        [750,300],
                        [750,200],
                        [750,100],
                        [300,250]]
        }
      },
      bids: [{
         bidder: 'sspBC',
         params: {
                    id: '003',
                    siteId: '235860',
                    test: 1          
         }
       }]
   }
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer: wojciech.bialy@grupawp.pl
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

File: modules/sspBCAdapter.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+            const bid = {
+              requestId: bidRequest.bidId,
+              creativeId: serverBid.crid || 'bc' + Math.floor(1E10 * Math.random()),
+              cpm: bidCpm || serverBid.price,
+              currency: response.cur,
+              ttl: serverBid.exp || 300,
+              width: serverBid.w,
+              height: serverBid.h,
+              bidderCode: BIDDER_CODE,
+              mediaType: 'banner',
```
[the remark this answers is not in the document]
Comment (patmmccann) [reply modules/sspBCAdapter.js:275]: see http://prebid.org/dev-docs/bidder-adaptor.html

[the remark this answers is not in the document]
Comment (wojciech-bialy-wpm) [reply modules/sspBCAdapter.js:275]: Understood! I've looked into it, and while our JSON endpoint does not return bid metadata, it should be available server-side (as underlying logic is using openRTB protocol).

I've created a change request, will update adapter & pull request once it's done.

**Comment by mike-chowla** [comment]
I'm not getting a test bid back

Request URL: https://ssp.wp.pl/bidder/?cs=true&bdver=4.4&pbver=3.27.0-pre&inver=0
Request Body
`{"id":"a068e9b1-69a3-458c-a556-bbd46d66ab64","site":{"id":"235911","page":"somesite.pl/somepage.html","domain":"somesite.pl","ref":"http://localhost:9999/integrationExamples/gpt/"},"imp":[{"id":"006","banner":{"format":[{"w":300,"h":250},{"w":300,"h":600}],"id":"2935dffd2e0f14"},"tagid":"006"}],"tmax":250,"user":{"data":[{"id":"12","name":"NetInfo","segment":[{"name":"CH-Ect","value":"4g"},{"name":"CH-Rtt","value":"100"},{"name":"CH-Downlink","value":"5.7"},{"name":"CH-DeviceMemory","value":"8"},{"name":"CH-Dpr","value":"1"},{"name":"CH-ViewportWidth","value":"1787"}]}]},"regs":{}}`

File: modules/sspBCAdapter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (wojciech-bialy-wpm) [reply modules/sspBCAdapter.js:278]: Adapter & test have been updated, bids now have meta.advertiserDomains property:
`              meta: {
                advertiserDomains: serverBid.adomain,
              },
`

**Comment by wojciech-bialy-wpm** [comment]
"Not getting a test bid back" - this has been resolved. The was an issue with DSP not sending test ad. We're now enforcing test demand by adding test: 1 to bid parameters:
`
            "bids": [{
                "bidder": "sspBC",
                "params": {
                    "id": "003",
                    "siteId": "235860",
                    "test": 1
                }
            }],
`

I've created two simple test pages:
- one where prebid creative is rendered in a friendly frame: https://bdr.wpcdn.pl/tests/bidderv4/test-adapter_google.html
- and one where we're using Google Ad Manager, and Universal Creative in a safeframe: https://bdr.wpcdn.pl/tests/bidderv4/test-adapter_google.html

**Comment by mike-chowla** [comment]
I'm still not getting a bid back.  Here's the request body:

`{"id":"3b57cb24-323e-47c8-aefb-a71f215d4d33","site":{"id":"235860","page":"http://mike3.pubmatic.com:9999/integrationExamples/imac/hello_world.html?pbjs_debug=true","domain":"mike3.pubmatic.com"},"imp":[{"id":"003","banner":{"format":[{"w":300,"h":250},{"w":300,"h":600}],"id":"2728ae76cdde47"},"tagid":"003"}],"tmax":450,"user":{"data":[{"id":"12","name":"NetInfo","segment":[{"name":"CH-Ect","value":"4g"},{"name":"CH-Rtt","value":"50"},{"name":"CH-Downlink","value":"10"},{"name":"CH-Dpr","value":"2"},{"name":"CH-ViewportWidth","value":"1787"}]}]},"regs":{},"test":1}`

**Comment by wojciech-bialy-wpm** [comment]
@mike-chowla 

Regarding not getting a bid back: The adapter accepts gdpr-compliant requests only. It will not forward any requests that do not have consent for vendor BusinessClick (vendor Id 676), and purposes 1 & 3.
This, unfortunately, overrides any test settings.

In above case, gdpr data is not supplied, and therefore request is not sent to SSPs.

Could You append GDPR data to the request? If You want, You can use this javascript snippet, that will set up IAB v1 compliant cmp API:
https://bdr.wpcdn.pl/tests/static_cmp.js

We can change the logic (so that test requests will be sent regardless of gdpr status), but this might take some time.

**Comment by patmmccann** [comment]
Why don't you hard code a consent string into your unit tests

**Comment by wojciech-bialy-wpm** [comment]
@patmmccann Thank You, good idea!

I've added two more tests, that will perfrom xhr call to our json endpoint. One with test settings & hardcoded consent, one with same settings but no gdpr.

File: test/spec/modules/sspBCAdapter_spec.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    });
+  });
+
+  describe('Ajax tests', function () {
+    const { bids_test, bidRequestTest, bidRequestTestNoGDPR } = prepareTestData();
+    const requestSingleTest = spec.buildRequests(bids_test, bidRequestTest);
+    const requestSingleTestNoGDPR = spec.buildRequests(bids_test, bidRequestTestNoGDPR);
+
+    it('should return test ad when request is made with test:1 and gdpr data', function (done) {
+      const xhr = new XMLHttpRequest();
```
Comment (mike-chowla) [inline test/spec/modules/sspBCAdapter_spec.js:290]: There should be no XHR or network requests in your tests.  Our tests will **never** pass if we are reliant on 250 vendor endpoints to be working.  All responses should be mocked in the test frame.out

File: test/spec/modules/sspBCAdapter_spec.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+          const response = xhr.responseText ? JSON.parse(xhr.responseText) : {};
+          expect(response.seatbid).to.be.an('array');
+          done();
+        }
+      }
+      xhr.send(requestSingleTest.data);
+    });
+
+    it('should return no ads for request without gdpr, regardless of test settings', function (done) {
+      const xhr = new XMLHttpRequest();
```
Comment (mike-chowla) [inline test/spec/modules/sspBCAdapter_spec.js:303]: See above

Comment (wojciech-bialy-wpm) [reply test/spec/modules/sspBCAdapter_spec.js]: @mike-chowla I've removed ajax tests. I have left bidRequestTest and bidRequestTestNoGDPR, which should provide configs that 
- always receive one test ad (bidRequestTest)
- never receive test ad, due to the lack of GDDPR (bidRequestTestNoGDPR)

Could You use config provided in bidRequestTest and check if you're getting a bid response?

**Comment by patmmccann** [comment]
> @patmmccann Thank You, good idea!
> 
> I've added two more tests, that will perfrom xhr call to our json endpoint. One with test settings & hardcoded consent, one with same settings but no gdpr.

Sorry, Mike makes a good point, my mistake.

**Comment by wojciech-bialy-wpm** [comment]
sspbc-adapter branch has been updated

**Comment by wojciech-bialy-wpm** [comment]
@mike-chowla 

I've updated sspbc-adapter branch - it should now incorporate all commits from Prebid.js/master as of 24.8.2020 08:40 GMT

Is there anything else that needs to be updated?

**Pull request merged**